File: labs/dm/2/F.java
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Scanner;

/**
 * @author Saveliy Bakturin
 *
 * Don't write off, if you don't wanna be banned!
 */

public class F {
    public static void main(final String... args) {
        final Scanner in = new Scanner(System.in);
        final int n = in.nextInt();
        in.nextLine();
        final int[] input = new int[n];
        for (int i = 0; i < n; i++) {
            input[i] = in.nextInt();
        }
        in.close();
        final Map<Integer, String> alphabet = new LinkedHashMap<>();
        final Map<String, Integer> alphabetByString = new LinkedHashMap<>();
        for (int i = 0; i < 26; i++) {
            alphabet.put(i, String.valueOf((char) (i + 97)));
            alphabetByString.put(String.valueOf((char) (i + 97)), i);
        }
        int pointer = 26;
        final StringBuilder answer = new StringBuilder();
        answer.append(alphabet.get(input[0]));
        int prev_index = 0;
        alphabet.get(input[0]);
        lzw(n, input, alphabet, alphabetByString, pointer, answer, prev_index);
        System.out.println(answer);
    }

    private static void lzw(final int n, final int[] input, final Map<Integer, String> alphabet, final Map<String, Integer> alphabetByString, int pointer, StringBuilder answer, int prev_index) {
        for (int i = 1; i < n; i++) {
            if (alphabet.containsKey(input[i])) {
                answer.append(alphabet.get(input[i]));
                String s1 = alphabet.get(input[prev_index]);
                String s2 = alphabet.get(input[i]);
                for (int j = 0; j < s2.length(); j++) {
                    String s3 = s1 + s2.charAt(j);
                    if (alphabetByString.containsKey(s3)) {
                        s1 = s3;
                    } else {
                        alphabetByString.put(s3, pointer);
                        alphabet.put(pointer, s3);
                        prev_index = i;
                        pointer++;
                        break;
                    }
                }
            } else {
                String s1 = alphabet.get(input[prev_index]);
                String s2 = s1 + s1.charAt(0);
                answer.append(s2);
                alphabet.put(pointer, s2);
                alphabetByString.put(s2, pointer);
                pointer++;
                prev_index = i;
            }
        }
    }
}

File: algo/7/F.cpp
#include <cstdio>
#include <vector>
#include <set>
#include <algorithm>

/**
 * @author Saveliy Bakturin
 * <p>
 * Don't write off, if you don't wanna be banned!
 */

void dfs_ts(std::vector<std::vector<int>> const& edge, int const& v, std::vector<int> &ts, bool *visited) {
	visited[v] = true;
	for (int const& u : edge[v]) if (!visited[u]) dfs_ts(edge, u, ts, visited);
	ts.push_back(v);
}

void dfs_scc(std::vector<std::vector<int>> const& edge_reverse, int const& v, int const& c, std::vector<int> &comp) {
	comp[v] = c;
	for (int const& u : edge_reverse[v]) if (!comp[u]) dfs_scc(edge_reverse, u, c, comp);
}

int main() {
	std::size_t n, m;
	std::scanf("%zu %zu", &n, &m);
	bool *visited = new bool[n + 1]();
	std::vector<std::vector<int>> edge(n + 1, std::vector<int>());
	std::vector<std::vector<int>> edge_reverse(n + 1, std::vector<int>());
	std::vector<int> comp(n + 1);
	std::vector<int> ts;
	std::set<std::pair<int, int>> answer;
	int c = 1;
	for (std::size_t i = 0; i < m; i++) {
		int u, v;
		std::scanf("%i %i", &u, &v);
		edge[u].push_back(v);
		edge_reverse[v].push_back(u);
	}
	for (int v = 1; v < n + 1; v++) if (!visited[v]) dfs_ts(edge, v, ts, visited);
	std::reverse(ts.begin(), ts.end());
	for (int const& v : ts) if (!comp[v]) dfs_scc(edge_reverse, v, c++, comp);
	for (int u = 1; u < n + 1; u++) {
		for (int v : edge[u]) if (comp[u] != comp[v]) answer.emplace(comp[u], comp[v]);
	}
	std::printf("%zu\n", answer.size());
}
